students/brian_warn/session06/test_mailroom.py

```diff
@@ -0,0 +1,18 @@
+#!/Library/Frameworks/Python.framework/Versions/3.6/bin/python
+ # test_mailroom.py
+
+import mailroom_from_chb as mr
+
+import pytest
+
+
+def test_get_donor_db():
+    result = mr.get_donor_db()
+    assert result == {'william gates iii': ("William Gates III", [653772.32, 12.17]),
+        'jeff bezos': ("Jeff Bezos", [877.33]),
+        'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
+        'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),}
+
+def test_list_donors():
```

Comment:
@PythonCHB @codedragon This fails with `NameError: name 'donor_db' is not defined` error.  I don't know how to test for functions that get content outside of the function itself.

```
___________________________________________________________________________ test_list_donors ___________________________________________________________________________

    def test_list_donors():
>       result = mr.list_donors()

test_mailroom.py:17: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def list_donors():
        """
        Create a list of the donors as a string, so they can be printed
    
        Not calling print from here makes it more flexible and easier to
        test
        """
        listing = ["Donor list:"]
>       for donor in donor_db.values():
E       NameError: name 'donor_db' is not defined

mailroom_from_chb.py:39: NameError
```

Reply:
Brian -- you are running into an anticipated issue!

Remember that I said something like "if your code is hard to test, that probably means it could use some refactoring!

In fact, I originally wrote mailroom without tests, and when I went to add the tests, I ended up having to refactor it a bit to make it more testable.

What you are running into now is one of the reasons why people generally frown on using "globals" -- it means that a function can't run properly outside of it's original context.

(it also means that you can't easily tell where things might get altered.)

I'll take a look at what you've got and see what I can suggest for refactoring -- either the code or your tests :-)

Reply:
OK -- found the solution.

Since the code IS in fact structured around a global donor_db, then you need to make sure it exists when you run your tests.

Note in the code, that the donor_db is created by a function --this is so that it would be fairly easy to trade-out code that, for instance, read the db from a file, or ...

So in the script, the donor_db is loaded into a global variable in the if __name__ block:

```
if __name__ == "__main__":
    donor_db = get_donor_db()
```

but, as you've noticed, that code doesn't get run when you run your tests!

So you need to run it by hand in your test file:

```
import mailroom2 as mailroom
mailroom.donor_db = mailroom.get_donor_db()
```
that creates the database, and sets it to a module-global variable.

Pretty wierd that you can set a module variable from outside the module -- but you can!

Ideally, that code would be in a "Fixture" -- but I don't want to complicate things for you too much, and using a global like that makes it hard to isolate anyway.